### ISSUE TITLE
[Question] Confused about the proper setup for a Kustomize project

### ISSUE THREAD

**Issue by chrispytoes** [kubernetes-sigs/kustomize#3971]
I am somewhat new to Kubernetes, and I'm setting up a Kustomize project for all of my configs. As a programmer, I'm a little confused about the documentation and the flat project structure it seems to encourage. I tried nesting one Kustomization in another, something like this:

    cluster
      app1
        deployments
          deployment1.yaml
          deployment2.yaml
          kustomization.yaml
            | resources:
            | - deployment1.yaml
            | - deployment2.yaml
        services
          service1.yaml
          service2.yaml
          service3.yaml
          kustomization.yaml
            | resources:
            | - service1.yaml
            | - service2.yaml
            | - service3.yaml
        kustomization.yaml
          | resources:
          | - ./deployments
          | - ./services
      kustomization.yaml
        | resources:
        | - ./app1

NOTE: I am not trying to depict separate development and production deployments with this example. I understand that's what the inheritance system is for. I simply want to nest/group my kustomizations arbitrarily, purely for organizational purposes. I expect this cluster to grow rather large over time. Keeping all of my apps alongside my common bases, transformers, and pairs right at the same directory level seems like it would get pretty ugly pretty quick. However when I try to apply an example like this, I get an error saying that a resource must be a file.

I think whole the base/overlay system is neat and all, but it's only half the picture. Not being able to nest subfolders is a huge killer for me. This also means that you need to run a different command depending on which `kustomization.yaml` you want to apply. You can't simply have one root `kustomization.yaml` that calls out to all the others and applies everything.

As a programmer, this feels like common sense to me, so I'm wondering if there's something fundamental I'm not understanding here.

**Comment by natasha41575** [comment]
I think we need some more information about how you are using kustomize.

I have the following tree structure: 
```
cluster
├── app1
│   ├── deployments
│   │   ├── kustomization.yaml
│   │   └── deployment.yaml
│   └── kustomization.yaml
└── kustomization.yaml
```

where `cluster/kustomization.yam` is:
```
resources:
  - ./app1
```
 
`cluster/app1/kustomization.yaml`:
```
resources:
  - ./deployments
```

`cluster/app1/deployments/kustomization.yaml`:
```
resources:
  ./ deployment.yaml
```

which (correct me if I'm wrong) seems to be the scenario that you're describing. Running `kustomize build cluster` works for me. Would you mind providing the full contents of your kustomization files or some other information to help us reproduce your issue?

**Comment by natasha41575** [comment]
Another question: which version of kustomize are you using?
If you are using `kubectl apply -k`, which version of kubectl are you using?

**Comment by chrispytoes** [comment]
@natasha41575 You are correct, that is my exact scenario. Using `kubectl apply -k` gives me the following error:

    error: rawResources failed to read Resources: Load from path ./ingress-nginx failed: './ingress-nginx' must be a file (got d='E:\projects\gatsby\wf-infrastructure\kustomize\ingress-nginx')

The kustomize.yaml it's complaining about just contains:
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
  - ./ingress-nginx
```
That ingress-nginx folder contains another kustomize.yaml.

Here's what the version about shows:
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:23:52Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:20:00Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by natasha41575** [comment]
The version of kustomize shipped with older kubectl versions (<= v.1.20) is very old. If you can, please use v.1.21.

If you are unable to upgrade your version of kubectl, you may have to use the old, deprecated `bases` field for directories and `resources` for files.

**Comment by chrispytoes** [comment]
@natasha41575 Interesting, I just installed `kubectl` for the first time about a month ago. I followed DigitalOcean's instructions so that's probably what directed me to the older version. I'll see if I can upgrade it.
